qiskit/algorithms/time_evolvers/variational/solvers/var_qte_linear_solver.py

```diff
@@ -0,0 +1,134 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Class for solving linear equations for Quantum Time Evolution."""
+from __future__ import annotations
+
+import copy
+from typing import Callable
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter
+from qiskit.quantum_info import SparsePauliOp
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+
+from .ode.assign_params import assign_parameters
+
+from ..variational_principles import VariationalPrinciple
+
+
+class VarQTELinearSolver:
+    """Class for solving linear equations for Quantum Time Evolution."""
+
+    def __init__(
+        self,
+        var_principle: VariationalPrinciple,
+        hamiltonian: BaseOperator,
+        ansatz: QuantumCircuit,
+        gradient_params: list[Parameter] | None = None,
+        t_param: Parameter | None = None,
+        lse_solver: Callable[[np.ndarray, np.ndarray], np.ndarray] | None = None,
+        imag_part_tol: float = 1e-7,
+    ) -> None:
+        """
+        Args:
+            var_principle: Variational Principle to be used.
+            hamiltonian: Operator used for Variational Quantum Time Evolution.
+            ansatz: Quantum state in the form of a parametrized quantum circuit.
+            gradient_params: List of parameters with respect to which gradients should be computed.
+                If ``None`` given, gradients w.r.t. all parameters will be computed.
+            t_param: Time parameter in case of a time-dependent Hamiltonian.
+            lse_solver: Linear system of equations solver callable. It accepts ``A`` and ``b`` to
+                solve ``Ax=b`` and returns ``x``. If ``None``, the default ``np.linalg.lstsq``
+                solver is used.
+            imag_part_tol: Allowed value of an imaginary part that can be neglected if no
+                imaginary part is expected.
+
+        Raises:
+            TypeError: If t_param is provided and Hamiltonian is not of type SparsePauliOp.
+        """
+        self._var_principle = var_principle
+        self._hamiltonian = hamiltonian
+        self._ansatz = ansatz
+        self._gradient_params = gradient_params
+        self._bind_params = gradient_params
+        self._time_param = t_param
+        self.lse_solver = lse_solver
```

Comment:
This appears to be a public instance var - do w need this - if so we need to doc this via Attributes